ompi/errhandler/errcode.c

```diff
@@ -216,6 +217,7 @@ int ompi_mpi_errcode_init (void)
     CONSTRUCT_ERRCODE( ompi_err_rma_shared, MPI_ERR_RMA_SHARED, "MPI_ERR_RMA_SHARED: Memory cannot be shared" );
     CONSTRUCT_ERRCODE( ompi_t_err_invalid, MPI_T_ERR_INVALID, "MPI_T_ERR_INVALID: Invalid use of the interface or bad parameter value(s)" );
     CONSTRUCT_ERRCODE( ompi_t_err_invalid_name, MPI_T_ERR_INVALID_NAME, "MPI_T_ERR_INVALID_NAME: The variable or category name is invalid" );
+    CONSTRUCT_ERRCODE( ompi_err_rma_nb_pending, MPI_ERR_RMA_NB_PENDING, "MPI_ERR_RMA_NB_PENDING: rma nonblocking pending operation" );
```

Comment:
Hmm, the "error" of RMA pending should probably *never* be given to the user code, so we should not need to have a new `MPI_ERR_...` code; the new `OMPI_ERR_...` code is needed for internal use only.

If we exposed a user-facing API for this feature, such as `MPI_WIN_ILOCK`, it would return `MPI_SUCCESS` to the user even when the lock had not been successful. A subsequent `MPI_WAIT` for the lock would block execution until it could also return `MPI_SUCCESS`, meaning that the lock had eventually been successful.

OTOH, if we exposed a user-facing API for this feature, such as `MPI_WIN_TRYLOCK`, *that* might return `MPI_SUCCESS` when the lock attempt was successful or `MPI_ERR_RMA_NB_PENDING` (or similar) when the lock attempt was not successful. It would probably be better API design to have such a `TRY` function take a `flag` parameter, which would set to `true` if the lock attempt was successful and `false` otherwise. That reserves the error code for actual errors -  `MPI_ERR_RMA_NB_PENDING` is not an actual error, it is a "not yet" indication; it is just like `MPI_TEST` setting `flag` to `false` to mean "not yet; try/test again". With that API design, the following is possible:

``` C
int MPI_WIN_LOCK (...) {
    int ret = MPI_SUCCESS;
    int flag = 0;
    while (MPI_SUCCESS==ret && 0==flag) {
        ret = MPI_WIN_TRYLOCK (..., &flag);
        OMPI_do_progress();
    }
    return ret;
}

int MPI_WIN_TRYLOCK (..., int *flag) {
    int ret = MPI_SUCCESS;
    *flag = 0;
    ret = ompi_trylock_internal (...);
    if (ompi_err_rma_nb_pending==ret)
        ret = MPI_SUCCESS;
    else if (MPI_SUCCESS==ret)
        *flag = 1;
    return ret;
}
```

Reply:
We can take a look at this in the future if it is interesting for the MPI forum.

Reply:
We will not expose the 'ERR_NB_PENDING' to the user for the time being.